Log out user account after account deletion success

charliescheer opened the pull request:
### Fix
Per some discussions with in app account deletion for MacOS, we decided that it would be best to log out a user once the account deletion is complete, instead of simply displaying the onboarding.  We did this on macOS to deal with some windowing bugs, which were not apparent on iOS, but could lead to being able to navigate the app after an account was deleted making it seem like the deletion had failed.

By logging the user out the local data is removed and there is no way to navigate the app. 
### Test
1.  Create an account to delete
2. delete that account.
3. open simplenote iOS back up, notice that with in a moment or two the onboarding VC will appear.
4. send simplenote to the background and open it again
5. You will see the slide out menu for a short moment (this is in part because of the restoration state and I do not think is avoidable) but it can't be interacted with and the onboarding vc should appear quickly

### Review
***(Required)*** Add instructions for reviewers.  For example:
> Only one developer are required to review these changes, but anyone can perform the review.

### Release

> These changes do not require release notes.

charliescheer committed:
Added logic to logout user when deletion request completed

charliescheer committed:
Changed removal of credentials to use authenticator.reset

eshurakov commented on Simplenote/SPAppDelegate+Extensions.swift:
```diff
@@ -278,6 +278,7 @@ extension SPAppDelegate: SimperiumDelegate {
 
     public func simperium(_ simperium: Simperium, didFailWithError error: Error) {
         SPTracker.refreshMetadataForAnonymousUser()
+        logOutIfAccountDeletionRequested()
```
We should only handle `SPSimperiumErrorsInvalidToken` error here as that's the only case we're interested in

jleandroperez replied:
+1 to this one 😄

eshurakov commented on Simplenote/SPAppDelegate+Extensions.swift:
```diff
@@ -408,16 +409,25 @@ extension SPAppDelegate {
+        guard let deletionController = accountDeletionController,
+              deletionController.hasValidDeletionRequest,
+              let user = simperium.user,
+              user.authenticated() else {
```
`user.authenticated()` will be false because we come here when user token is invalid, which means user is not authenticated. 
We don't need this condition here and can safely remove it.

eshurakov replied:
After I removed `user.authenticated()` and tested the account deletion flow, I found another issue.

Calling `[self.simperium signOutAndRemoveLocalData:YES completion:^{}];` doesn't do anything because the first line there is checking if user is authenticated. It will return `false` in our case because the token is not valid and `singOut` won't do anything because of that. This means that all the data is still stored in the core data and we failed to remove it.

https://github.com/Simperium/simperium-ios/blob/develop/Simperium/Simperium.m#L589
```
- (void)signOutAndRemoveLocalData:(BOOL)remove completion:(SimperiumSignoutCompletion)completion {
    
    // Don't proceed, if the user isn't logged in
    if (!self.user.authenticated) {
        return;
    }

   ...
```

The issue here is that simperium checks for `self.user.authenticated` which can be false, but we will still have the user object and credentials stored in the keychain and on the next app launch Simperium will try to use the token from the keychain. The check is quite misleading.

I suggest that we change ` if (!self.user.authenticated)` to ` if (!self.user)` or even fully remove the line. This should allow us to fully clean the state even if the user is not authenticated.

@jleandroperez WDYT?

jleandroperez replied:
Simperium [updated here](https://github.com/Simperium/simperium-ios/pull/622). TY Evgeny!! 

+1 yes let's just check if there's a valid deletion request pending.
Plus: we should point to the new Simperium branch (I'll make sure to release a new version once we know it's good!)

eshurakov commented on Simplenote/SPConstants.h:
```diff
@@ -23,6 +23,7 @@ extern NSString *const kAutomatticAnalyticLearnMoreURL;
 
 extern NSString *const kShareExtensionGroupName;
 extern NSString *const kSimplenoteWPServiceName;
+extern NSString *const SPUsername;
```
This is not used :-)

jleandroperez commented on Simplenote/SPAppDelegate+Extensions.swift:
```diff
@@ -408,16 +409,25 @@ extension SPAppDelegate {
-        guard let deletionController = accountDeletionController else {
+    func authenticateSimperiumIfAccountDeletionRequested() {
+        guard let deletionController = accountDeletionController,
+              deletionController.hasValidDeletionRequest else {
```
You can collapse this into a single check:

```
guard accountDeletionController?.hasValidDeletionRequest == true else { //...
```

jleandroperez commented on Simplenote/SPAppDelegate.m:
```diff
@@ -325,6 +327,12 @@ - (void)logoutAndReset:(id)sender
 
+-(void)removeStoredCredentialsFromKeychain
+{
+    [self.simperium.authenticator reset];
```
We should prob. also drop this invocation

jleandroperez reviewed (commented):
Thank you gentleman 😄

charliescheer committed:
Removed app delegate call to log user out

charliescheer committed:
Added check to log out only if the correct error is thrown

charliescheer committed:
Fixed a bug that would cause the onboarding vc to load/dismiss forever

charliescheer committed:
Fixed a bug the onboarding vc was dismissed/restored after deletion

charliescheer commented:
Alright, tested the simperium changes, and that does work.  It created a new issue where the onboarding screen was loaded for a moment then it is dismissed, and then it is displayed again.  This happens because of simperium.authenticationDidFail schedules an onboarding VC to appear and logout calls dismissAllModals and then also calls authenticateIfNecessary which schedules a new onboarding VC.  I removed the call to dismiss  all modals to fix the issue.

Also I added a change that fixes #1393  (adding that in so it will auto close when this is merged)

To test these changes with the simperium changes in the SNiOS Pod file comment out 
`pod 'Simperium', '1.8.0'`

and add

`pod 'Simperium', :git => 'https://github.com/Simperium/simperium-ios.git', :branch => 'issue/drops-unneeded-logout-check'`

jleandroperez commented on Simplenote/SPAppDelegate+Extensions.swift:
```diff
@@ -278,6 +278,18 @@ extension SPAppDelegate: SimperiumDelegate {
+        switch simperiumError {
+        case .invalidToken:
+            logOutIfAccountDeletionRequested()
+            break
```
No `break` needed here

jleandroperez commented on Simplenote/SPAppDelegate.m:
```diff
@@ -315,12 +315,11 @@ - (void)logoutAndReset:(id)sender
 			// hide sidebar of notelist
             [self.sidebarViewController hideSidebarWithAnimation:NO];
-			
-			[self dismissAllModalsAnimated:YES completion:^{
```
I believe `dismissAllModals` was introduced to dismiss the spinner (and we should probably check on iPad too).

How can I reproduce the loop-y behavior? (Already tried reverting the Simperium change, and dropping this). I think we should probably either keep the `dismissAllModals`, or analyze very closely all of the things that get dismissed by this call.

Commit:
Reverts App level TokenInvalid handler

Commit:
Fixes weird alignment

Commit:
Revert "Fixes weird alignment"

This reverts commit 7f025c26e1a0f26fe3f1bae2bc0995853c0752f8.

Commit:
Revert "Reverts App level TokenInvalid handler"

This reverts commit 0e0a7bfa5575fe682cea04da35df7c29ed9c7651.

Commit:
Wires latest Simperium

Commit:
Restores dismissAllModalsAnimated

Commit:
Adds missing Spinner dismiss

Commit:
Relocates clearRequestToken

eshurakov reviewed (approved):
Worked great for me!

charliescheer commented:
I checked it out as well and it looks good to me @jleandroperez

Commit:
Updates Simperium to Mk 1.9.0

Commit:
Updates Simperium dependency

jleandroperez reviewed (approved):
:shipit: sir!
